<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>2. Add a dataset &mdash; Epigenome Data Browse-R 0.1 beta documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1 beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Epigenome Data Browse-R 0.1 beta documentation" href="index.html" />
    <link rel="next" title="5. Navigating the Browse-R" href="basic_view.html" />
    <link rel="prev" title="1. Install instructions" href="install.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="basic_view.html" title="5. Navigating the Browse-R"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="install.html" title="1. Install instructions"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">Epigenome Data Browse-R 0.1 beta documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="toctree-wrapper compound" id="add-dataset">
<ul class="simple">
</ul>
</div>
<div class="section" id="add-a-dataset">
<h1>2. Add a dataset<a class="headerlink" href="#add-a-dataset" title="Permalink to this headline">¶</a></h1>
<div class="section" id="basic-steps">
<h2>2.1. Basic steps<a class="headerlink" href="#basic-steps" title="Permalink to this headline">¶</a></h2>
<p>Follow these steps to add a new dataset to the EDB. As an example, we use the demo BS-seq dataset from Lister, Mukamel et al. (2013). Science, which charts DNA methylation over human postnatal brain development.</p>
<p>Let us assume that EDB datasets are all located at <tt class="docutils literal"><span class="pre">&lt;dataRoot&gt;</span></tt>.</p>
<p><strong>Step 1: dataset config:</strong> Create a <a class="reference internal" href="#add-data-config"><em>master config file</em></a> for the dataset.</p>
<p><strong>Step 2: phenotype matrix:</strong> Add a <a class="reference internal" href="#add-data-pheno"><em>table specifying sample phenotypes/metadata properties</em></a>, and links to the sample-wise data</p>
<p><strong>Step 3: data:</strong> Assemble the data files and any required index files (e.g. <em>tbi</em> for tabix files).</p>
<p><strong>Step 4: group order:</strong> Add a file to specify the <a class="reference internal" href="#add-data-grouping"><em>ordering of categorical variables</em></a></p>
</div>
<div class="section" id="example-directory-structure-for-datasets">
<h2>2.2. Example directory structure for datasets<a class="headerlink" href="#example-directory-structure-for-datasets" title="Permalink to this headline">¶</a></h2>
<p id="add-data-exampledir">This is what a simple project directory structure should look like when you are done. The example below assumes that all datasets and associated config files are located at <tt class="docutils literal"><span class="pre">&lt;dataRoot&gt;</span></tt>. Coloured arrows show how the master config file references the phenotype table and group order table</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/dataset_anatomy.png"><img alt="Conceptual view of project directory" src="_images/dataset_anatomy.png" style="width: 900px;" /></a>
</div>
<p>In this example we have two datasets, one showing epigenetic dynamics in mouse brain development and the other, for human. Here, <tt class="code docutils literal"><span class="pre">config_location.txt</span></tt> points to a directory which also contains all browser-related metadata and the data files themselves:</p>
<div class="highlight-python"><div class="highlight"><pre>&lt;config_location_points_here&gt;
         |
         |---- mouseBrainDev_config.txt *symlink*
         |---- humanBrainDNAMethylome_config.txt *symlink*
         |---- mouseBrainDev/
                        |---- config.txt
                        |---- pheno.txt
                        |---- group_order.txt
                        |---- data
                                  |---- GSM123456_mm9.bw
                                  |---- GSM654321_mm9.bw
                                  | ...
                                  |---- GSM9999_mm9.bw
         |---- humanBrainDNAMethylome/
                        |---- config.txt
                        |---- pheno.txt
                        |---- group_order.txt
                        |---- data
                                  |----
                                  |---- PCW4.bw
                                  |---- midGestation_Cortex.bw
                                  | ...
                                  |---- PostPuberty_AnteriorCingulateCortex.bw
     |----anno
</pre></div>
</div>
</div>
<div class="section" id="dataset-config-file">
<h2>2.3. Dataset config file<a class="headerlink" href="#dataset-config-file" title="Permalink to this headline">¶</a></h2>
<p id="add-data-config">The dataset config file contains general metadata for the dataset. It is a tab-delimited file with two columns: a key (controlled word for config parameter) and value.</p>
<div class="section" id="required-fields">
<h3>2.3.1. Required fields<a class="headerlink" href="#required-fields" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">name</span></tt>: Dataset name; human-readable (~70 char)</li>
<li><tt class="docutils literal"><span class="pre">description</span></tt>: Brief description of dataset; human-readable (~170 char)</li>
<li><tt class="docutils literal"><span class="pre">platformName</span></tt>: Short description of platform on which data was generated; human-readable (~30 char)</li>
<li><tt class="docutils literal"><span class="pre">datasetConfig</span></tt>: Path to <a class="reference internal" href="#add-data-pheno"><em>sample phenotype table</em></a></li>
<li><tt class="docutils literal"><span class="pre">groupCols</span></tt>: Column names of <cite>sample phenotype table &lt;add-data-pheno&gt;</cite> that EDB should allow samples to be grouped by; comma-separated list, case-sensitive</li>
<li><tt class="docutils literal"><span class="pre">groupOrder</span></tt>: Path to <em class="xref std std-ref">group order file</em></li>
<li><tt class="docutils literal"><span class="pre">defaultGroup</span></tt>: Name of group that serves as the default value in the <em>Group by</em> dropdown box of EDB interface; must be a column name of the <cite>sample phenotype table &lt;add-data-pheno&gt;</cite>, case-sensitive</li>
<li><tt class="docutils literal"><span class="pre">chromSizes</span></tt>: Path to text file containing genome sequence sizes. This file should be the output of <cite>fetchChromSizes</cite> from Kent utilities at UCSC.</li>
<li><tt class="docutils literal"><span class="pre">genomeName</span></tt>: Name of genome build ; must correspond to a <cite>UCSC release name &lt;https://genome.ucsc.edu/FAQ/FAQreleases.html&gt;</cite></li>
<li><tt class="docutils literal"><span class="pre">ideoFile</span></tt>: Path to text file corresponding to UCSC <tt class="docutils literal"><span class="pre">cytoBandIdeo</span></tt> table. <strong>Header?</strong></li>
<li><tt class="docutils literal"><span class="pre">annoConfig</span></tt>: Path to config file for genome annotation. <strong>TODO: Add link</strong></li>
<li><tt class="docutils literal"><span class="pre">datatype</span></tt>: Datatype. Currently, one of <tt class="docutils literal"><span class="pre">{</span> <span class="pre">bigwig</span> <span class="pre">,</span> <span class="pre">BSseq</span> <span class="pre">}</span></tt>. Indicates how data should be processed; most single continuous traces (e.g. coverage, tiling microarrays) can use the &#8216;bigwig&#8217; setting.</li>
</ul>
<p>Separately, there may be datatype-specific config parameters.</p>
</div>
<div class="section" id="bigwig">
<h3>2.3.2. bigwig<a class="headerlink" href="#bigwig" title="Permalink to this headline">¶</a></h3>
<p>This datatype has no required parameters.</p>
</div>
<div class="section" id="bsseq">
<h3>2.3.3. BSseq<a class="headerlink" href="#bsseq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">BSseq__COV_POS</span></tt>: Column index of tabix file which contains base coverage</li>
<li><tt class="docutils literal"><span class="pre">BSseq__M_POS</span></tt>: Column index of tabix file which contains number of methylated reads at that base (&#8220;M&#8221; read count)</li>
</ul>
</div>
<div class="section" id="example-config-file">
<h3>2.3.4. Example config file<a class="headerlink" href="#example-config-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre># REQUIRED: MAIN config flags --- must be in all datasets
# ----------------------------------------------
name    Human brain development methylome (base-resolution BS-seq)
description DNA modifications assayed by whole-genome bisulfite sequencing over postnatal development. Samples from Lister, Mukamel et al. (2013). Science. 341. GEO:GSE47966.
platformName    BS-seq; Illumina HiSeq 2000.
#Path to dataset config file (sample file source location, group names, etc.,)
datasetConfig   /&lt;dataRoot&gt;/ListerEcker_2013/pheno_ListerEcker2013.txt
groupCols   Age_Group,Cell_Type,Sex
groupOrder  /&lt;dataRoot&gt;/ListerEcker_2013/group_order.txt
defaultGroup    Age_Group
# path to chromosome sizes  - output of Kent utilities fetchChromSizes
chromSizes  /&lt;dataRoot&gt;/anno/hg19/hg19.chrom.sizes
genomeName  hg19
ideoFile    /&lt;dataRoot&gt;/anno/hg19/cytoBandIdeo.txt
annoConfig  /&lt;dataRoot&gt;/anno/hg19/anno_config.txt
datatype    BSseq
ylabel  % methylation
#---------------------------------------------
# OPTIONAL: DATATYPE-specific; flags must have prefix &lt;datatype&gt;__
# Leave section blank if there are no flags.
#---------------------------------------------
BSseq__COV_POS  6
BSseq__M_POS    5
</pre></div>
</div>
<p>EDB requires that all dataset config files be located at the path specified in the special <tt class="docutils literal"><span class="pre">config_location.txt</span></tt> file. Recall that this file is at <tt class="docutils literal"><span class="pre">/path/to/EDB/code/config_location.txt</span></tt>.</p>
<p><a class="reference internal" href="#add-dataset"><em>Back to top</em></a></p>
</div>
</div>
<div class="section" id="sample-phenotype-table">
<h2>2.4. Sample phenotype table<a class="headerlink" href="#sample-phenotype-table" title="Permalink to this headline">¶</a></h2>
<p id="add-data-pheno">This tab-delimited file contains sample-wise metadata, including locations of data files. Each row should contain data for one sample, and each column should contain a unique type of metadata. Column order is unimportant to the browser.
The browser expects the following columns, named exactly in this way:</p>
<div class="section" id="required-columns">
<h3>2.4.1. Required columns<a class="headerlink" href="#required-columns" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>sampleName:</li>
<li>bigDataURL: Location of data file</li>
<li>all grouping columns as described in the <a class="reference internal" href="#add-data-grouping"><em>grouping order</em></a> file.</li>
</ul>
</div>
<div class="section" id="example-sample-phenotype-table">
<h3>2.4.2. Example sample phenotype table<a class="headerlink" href="#example-sample-phenotype-table" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>sampleName  Age_Group   Cell_Type   Bank_ID Sex bigDataURL
Human_35_do Early Childhood Tissue  UMB 759 M   /My_Shiny_DataDir/data/GSM116700
4_allC.MethylC-Seq_hs_mfg_35do.chr1.txt.gz
Human_2_yr  Early Childhood Tissue  UMB 5180    M   /My_Shiny_DataDir/data/GSM1167005_allC.MethylC-Seq_hs_mfg_2yr.chr1.txt.gz
Human_5_yr  Early Childhood Tissue  UMB 1185    M   /My_Shiny_DataDir/data/GSM1166274_allC.MethylC-Seq_hs_mfg_5yr.chr1.txt.gz
Human_12_yr Tween   Tissue  UMB 616 M   /My_Shiny_DataDir/data/GSM1164630_allC.MethylC-Seq_hs_mfg_12yr.chr1.txt.gz
Human_16_yr Adolescence Tissue  UMB 1158    M   /My_Shiny_DataDir/data/GSM1164631_allC.MethylC-Seq_hs_mfg_16yr.chr1.txt.gz
Human_25_yr Adult   Tissue  UMB 1829    M   /My_Shiny_DataDir/data/GSM1164632_allC.MethylC-Seq_hs_mfg_25yr.chr1.txt.gz
Hs 55 yr tissue Adult   Tissue  UMB 797 M   /My_Shiny_DataDir/data/GSM1173775_allC.MethylC-Seq_hs_fc_male_55yr_tissue.chr1.txt.gz
Hs 55 yr NeuN+  Adult   Neurons UMB 797 M   /My_Shiny_DataDir/data/GSM1173776_allC.MethylC-Seq_hs_fc_male_55yr_NeuN_pos.chr1.txt.gz
Hs 55 yr NeuN-  Adult   Non-Neurons UMB 797 M   /My_Shiny_DataDir/data/GSM1173777_allC.MethylC-Seq_hs_fc_male_55yr_NeuN_neg.chr1.txt.gz
HUES6   Immortal    ESC -   F   /My_Shiny_DataDir/data/GSM1173778_allC.MethylC-Seq_hs_hues6.chr1.txt.gz
</pre></div>
</div>
</div>
</div>
<div class="section" id="groups-and-grouping-order">
<h2>2.5. Groups and grouping order<a class="headerlink" href="#groups-and-grouping-order" title="Permalink to this headline">¶</a></h2>
<p id="add-data-grouping">The <tt class="code docutils literal"><span class="pre">group_order.txt</span></tt> file is a tab-delimited file containing a table of two columns:
#. groupID: Group name, must match a column name in the phenotype matrix
#. groupOrder: Order in which group members must be shown. Comma-separated collection of values. <strong>All values for a given group must be specified here. The browser will return an error if any additional group members are found in the phenotype table but are not listed here.</strong></p>
<p>In addition to these groups, the browser allows a non-grouping option - i.e. viewing sample-specific data - with &#8220;Grouping: (none)&#8221;.</p>
<p>As an example:</p>
<div class="highlight-python"><div class="highlight"><pre>groupID groupOrder
Tissue  Brain,Sperm
Diagnosis       Control,Schizophrenia,Bipolar disorder
TimeOfSampling  Before_Treatment,During_Treatment,After_Treatment
</pre></div>
</div>
<p>For this dataset, the browser would show 4 grouping options: Tissue, Diagnosis, TimeOfSampling, (none).</p>
<blockquote>
<div><a class="reference internal" href="#add-dataset"><em>Back to top</em></a></div></blockquote>
</div>
</div>
<div class="section" id="adding-demo-datasets">
<h1>3. Adding demo datasets<a class="headerlink" href="#adding-demo-datasets" title="Permalink to this headline">¶</a></h1>
<p><em>This section is added for completeness. Demo datasets have not yet been made publicly available. - SP 5 Sep 2014.</em></p>
<p>In this example, the demo data and config files are contained in a directory named &#8220;Shiny_BrowseR_Github_data&#8221;.</p>
<p>From the source machine, rsync the datasets to the machine where the EDB instance will be hsoted. The data directory should be in a location readable by the user shiny, which will run the EDB app.</p>
<div class="highlight-none"><div class="highlight"><pre>rsync -avL --progress Shiny_BrowseR_Github_data $DEST_MACHINE:/.
</pre></div>
</div>
<p>On the destination machine, we update the path in the metadata files to reflect the path on our new machine. <strong>Replace the substitution sed command with one relevant to your source and destination paths.</strong></p>
<p>Below we update the paths in the data dirs as well as the annotation directory (anno/hg19).</p>
<div class="highlight-none"><div class="highlight"><pre>cd /Shiny_BrowseR_Github_data
rm *_config.txt
cd mTAG_BrainSperm
sed -i &#39;s/\/src\/path\/dir/\/dest\/path/g&#39; *.txt
cd ../ListerEcker_2013
sed -i &#39;s/\/src\/path\/dir/\/dest\/path/g&#39; *.txt
cd ../anno/hg19
sed -i &#39;s/\/src\/path\/dir/\/dest\/path/g&#39; anno_config.txt
</pre></div>
</div>
<p>Create a symlink to dataset-specific config files in the data directory</p>
<p>We change <tt class="docutils literal"><span class="pre">/srv/shiny-server/EDB/config_location.txt</span></tt> to point to our new data directory path: <tt class="docutils literal"><span class="pre">/Shiny_BrowseR_Github_data</span></tt></p>
<p>At this point, refresh the EDB. If the &#8220;Choose dataset&#8221; dropdown box is populated, shiny can see the datasets.</p>
<div class="figure align-center">
<img alt="&quot;Choose dataset&quot; dropdown box populates, indicating EDB can see the data directory and read the config files." src="_images/edb_dataset_loads.png" />
</div>
<p>If not, stop here and check the following:</p>
<ul class="simple">
<li>Is the data root directory in a location with read permissions for user &#8220;shiny&#8221;?</li>
<li>Have the paths been correctly updated for all dataset directories?</li>
<li>Is config_location.txt pointing to the correct data directory?</li>
</ul>
</div>
<div class="section" id="add-annotation-sources">
<h1>4. Add annotation sources<a class="headerlink" href="#add-annotation-sources" title="Permalink to this headline">¶</a></h1>
<p>This is the directory structure for annotation sources:</p>
<div class="highlight-python"><div class="highlight"><pre>|------ hg19
                |------ cpgIslandExt.txt
                |------ cytoBandIdeo.txt
                |------ TxDb.Hsapiens.UCSC.hg19.refGene.sqlite
|------ mm9
                |------ cpgIslandExt.txt
                |------ LAD_NPC_mm9.txt
</pre></div>
</div>
<p>This is normal text again.</p>
<p><a class="reference internal" href="#add-dataset"><em>Back to top</em></a></p>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">2. Add a dataset</a><ul>
<li><a class="reference internal" href="#basic-steps">2.1. Basic steps</a></li>
<li><a class="reference internal" href="#example-directory-structure-for-datasets">2.2. Example directory structure for datasets</a></li>
<li><a class="reference internal" href="#dataset-config-file">2.3. Dataset config file</a><ul>
<li><a class="reference internal" href="#required-fields">2.3.1. Required fields</a></li>
<li><a class="reference internal" href="#bigwig">2.3.2. bigwig</a></li>
<li><a class="reference internal" href="#bsseq">2.3.3. BSseq</a></li>
<li><a class="reference internal" href="#example-config-file">2.3.4. Example config file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#sample-phenotype-table">2.4. Sample phenotype table</a><ul>
<li><a class="reference internal" href="#required-columns">2.4.1. Required columns</a></li>
<li><a class="reference internal" href="#example-sample-phenotype-table">2.4.2. Example sample phenotype table</a></li>
</ul>
</li>
<li><a class="reference internal" href="#groups-and-grouping-order">2.5. Groups and grouping order</a></li>
</ul>
</li>
<li><a class="reference internal" href="#adding-demo-datasets">3. Adding demo datasets</a></li>
<li><a class="reference internal" href="#add-annotation-sources">4. Add annotation sources</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="install.html"
                        title="previous chapter">1. Install instructions</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="basic_view.html"
                        title="next chapter">5. Navigating the Browse-R</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="basic_view.html" title="5. Navigating the Browse-R"
             >next</a> |</li>
        <li class="right" >
          <a href="install.html" title="1. Install instructions"
             >previous</a> |</li>
        <li><a href="index.html">Epigenome Data Browse-R 0.1 beta documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Shraddha Pai, Centre for Addiction and Mental Health.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>